### PR TITLE
Add an XML schema for documentation

### PR DESCRIPTION
**Pull request by Calinou**
`master` version of https://github.com/godotengine/godot/pull/59213.

This makes it easier to spot syntax errors when editing the class reference. The schema is referenced locally so validation can still work offline.

Tested with the [vscode-xml](https://github.com/redhat-developer/vscode-xml) Visual Studio Code add-on.

### CODE REVIEW

**Comment by follower** [comment]
> PS: Is there a way to forbid an element from containing _anything_ (even text)? This would be very nice to have for `<argument>` elements, as they can't have a description.

Based on a very cursory web search, It seems so, with a couple of variations:

 * <https://stackoverflow.com/questions/20751782/define-an-xml-element-that-must-be-empty-and-has-no-attributes>
 * <https://stackoverflow.com/questions/45077301/xsd-how-to-describe-an-element-so-it-can-have-no-value>
 * <https://www.w3schools.com/xml/schema_complex_empty.asp>
 * <https://www.ics.uci.edu/~alspaugh/cls/shr/xmlSchema.html#SimpleAndComplexContent>

Simplest option seems to be:

```xml
<xs:complexType/>
```

**Comment by Calinou** [comment]
I updated the pull request to add the XML schema mention in the root tag when using `--doctool`.

**Comment by bojidar-bg** [comment]
Seems like the schema path is wrong for `modules/*/doc_classes/*.xml`?

**Comment by Calinou** [comment]
@bojidar-bg Indeed… I don't know how we can elegantly fix this.

**Comment by bojidar-bg** [comment]
I'm not sure, but maybe it is possible get a path from `save_path` to `p_default_path.plus_file("../class.xsd")` in `DocData::save_classes`?

**Comment by Calinou** [comment]
@bojidar-bg Sounds good, but is there a method to "resolve" a path? Using `save_path.plus_file("../class.xsd"), I get `./doc/classes/../class.xsd` but I need `doc/class.xsd`.

**Comment by Calinou** [comment]
@bojidar-bg Thanks for the tip :slightly_smiling_face: I ended up using a different method as I didn't manage to do it using `String::plus_file()` or `String::to_path()`.

Also, I moved static checks to GitHub Actions.

File: doc/classes/@GlobalScope.xml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8" ?>
-<class name="@GlobalScope" version="4.0">
+<class name="@GlobalScope" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
```
Comment (aaronfranke) [inline doc/classes/@GlobalScope.xml:2]: Should this be `https://`?

**Review by aaronfranke — COMMENTED** [review]
How specifically does this make it easier to spot syntax errors when editing the class reference? AFAIK the XML is generated by Godot using `--doctool`, and shouldn't be changed manually except by adding text inside the tags.

**Comment by Calinou** [comment]
> How specifically does this make it easier to spot syntax errors when editing the class reference? AFAIK the XML is generated by Godot using `--doctool`, and shouldn't be changed manually except by adding text inside the tags.

This is mainly about spotting cases where users accidentally add new tags/attributes. This is mainly done by new contributors who are less familiar with the XML structure.

**Comment by pycbouh** [comment]
Last I checked, the doctool definitely has strict validation rules set via code. Thought I'm not sure if it fails when there are invalid tags or attributes. It definitely checks for the mandatory tags and attributes. We can probably enhance the DocTools class to fail or at least warn on malformed XMLs that contain unknown elements.

**Comment by pycbouh** [comment]
From helping some peeps on Discord I've figured the schema is probably way more useful than our ad-hoc checks we do with the doctool code. Namely, they report that adding arbitrary tags to the tutorials section, tags which are anything but `link`, breaks the description tags. Which is a weird bug, but also something a scheme could've highlighted, if we were to make it strict enough.

So overall, I think this would be beneficial, specifically to users who make custom modules and may be less well versed with how we structure our docs and all the limitations we have in place. I think we should do this.

@akien-mga what do you think?

**Comment by fire** [comment]
I support using existing xml tools to validate our xml variation. Xml validators allows tools to check our xml docs. We didn't need validators at first, but with the growth of Godot Engine scope and scale tool validators help.

**Comment by akien-mga** [comment]
I don't see code added to `DocTools` to actually write the XML schema to the header in new files, how does it work?

**Comment by akien-mga** [comment]
Some of your old commits referenced above in force push notifications haven't been GC'ed yet:

https://github.com/godotengine/godot/blob/f4dbc64abbf95f00f47e23ace284855a37004f29/editor/doc_data.cpp#L1068-L1073

**Review by akien-mga — APPROVED** [review]

**Comment by Calinou** [comment]
> Some XMLs have the wrong path: [godotengine/godot/runs/5187484781?check_suite_focus=true](https://github.com/godotengine/godot/runs/5187484781?check_suite_focus=true)

Fixed. I suppose this happens because I have the Mono module disabled, and the fallback text server is disabled when the advanced text server is enabled.

**Comment by akien-mga** [comment]
There's a style issue in the `DocData` changes.

**Comment by Calinou** [comment]
We could make the CI output quieter by using `xmllint`'s `--quiet` option, but this option isn't available in the version that ships with Ubuntu 18.04 :slightly_frowning_face: 

`--quiet` works with libxml 20912 from the Fedora 34 repositories.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**